src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java

```diff
@@ -54,10 +67,52 @@ public ScalaBuildServerService(BazelData bazelData, BazelRunner bazelRunner) {
     return CompletableFuture.completedFuture(new ScalaTestClassesResult(new ArrayList<>()));
   }
 
-  public CompletableFuture<ScalaMainClassesResult> buildTargetScalaMainClasses(
+  public Either<ResponseError, ScalaMainClassesResult> buildTargetScalaMainClasses(
       ScalaMainClassesParams scalaMainClassesParams) {
-    System.out.printf("DWH: Got buildTargetScalaMainClasses: %s%n", scalaMainClassesParams);
-    // TODO(illicitonion): Populate
-    return CompletableFuture.completedFuture(new ScalaMainClassesResult(new ArrayList<>()));
+
+    List<ScalaMainClassesItem> resultItems =
+        mainClassesItemTargetRulesResolver.getItemsForTargets(scalaMainClassesParams.getTargets());
+
+    ScalaMainClassesResult result = new ScalaMainClassesResult(resultItems);
+
+    return Either.forRight(result);
+  }
+
+  private ScalaMainClassesItem mapRuleToMainClassesItem(Build.Rule rule) {
+    BuildTargetIdentifier targetId = new BuildTargetIdentifier(rule.getName());
+    List<ScalaMainClass> mainClasses = collectMainClasses(rule);
+    return new ScalaMainClassesItem(targetId, mainClasses);
+  }
+
+  private List<ScalaMainClass> collectMainClasses(Build.Rule rule) {
+    Set<String> targetOptionsSet =
+        new LinkedHashSet<>(
+            collectAttributesFromStringListValues(rule, SCALA_COMPILER_OPTIONS_ATTR_NAME));
```

Comment:
I think I may have induced you in error with this, so I'm sorry about that :sweat_smile: . The `MainClass` request only receives options from the jvm, meaning the `jvm_flags` . The `scalacopts` field should be use on the `buildTarget/ScalacOptions`, not here